### PR TITLE
ARROW-16065: [FlightRPC][Docs] Improve Flight documentation

### PR DESCRIPTION
**Pull request by lidavidm**
This reworks the Flight protocol documentation and the C++ Flight documentation, and adds a brief Python documentation page. It also fixes some warnings and errors I found.

### CODE REVIEW

**Comment by lidavidm** [comment]
<details>
<summary>New protocol docs</summary>

image

</details>

<details>
<summary>New C++ docs</summary>

image

</details>

<details>
<summary>New Python docs</summary>

image

</details>

**Comment by lidavidm** [comment]
CC @davisusanibar we can base the Java documentation page on the Python/C++ ones here.

File: docs/source/cpp/build_system.rst [hunk 1]
```diff
@@ -176,7 +176,7 @@ text version of the IANA timezone database and add the Windows timezone mapping
 XML. To download, you can use the following batch script:
 
 .. literalinclude:: ../../../ci/appveyor-cpp-setup.bat
-   :language: cmd
+   :language: batch
```
Comment (wjones127) [inline docs/source/cpp/build_system.rst:179]: Thanks!

File: docs/source/cpp/flight.rst [hunk 2]
```diff
@@ -82,38 +94,78 @@ server stops.
...
 Using the Flight Client
 =======================
 
 To connect to a Flight service, create an instance of
 :class:`arrow::flight::FlightClient` by calling :func:`Connect
 <arrow::flight::FlightClient::Connect>`. This takes a Location and
-returns the client through an out parameter. To authenticate, call
-:func:`Authenticate <arrow::flight::FlightClient::Authenticate>` with
-the desired client authentication implementation.
+returns the client through an out parameter.
```
Comment (wjones127) [inline docs/source/cpp/flight.rst:103]: Should we be pointing users toward the `Result` variants of these functions instead given https://github.com/apache/arrow/pull/12719?

File: docs/source/cpp/flight.rst [hunk 3]
```diff
@@ -82,38 +94,78 @@ server stops.
...
+=========================
+
+When making a call, clients can optionally provide
+:class:`FlightCallOptions <arrow::flight::FlightCallOptions>`. This
+allows clients to set a timeout on calls or provide custom HTTP
+headers, among other features. Also, some objects returned by client
+RPC calls expose a ``Cancel`` method which allows terminating a call
+early.
+
+On the server side, timeouts are transparent. For cancellation, the
```
Comment (wjones127) [inline docs/source/cpp/flight.rst:119]: I find "transparent" isn't always straight foward in meaning. What do you think of this:

```suggestion
On the server side, no additional code is needed to implement timeouts. For cancellation, the
```

File: docs/source/format/Flight.rst [hunk 4]
```diff
@@ -29,73 +30,187 @@ either downloaded from or uploaded to another service. A set of
...
+   record batches.
+
+   If the server wishes to indicate that the data is on the local
+   server and not a different location, then it can return an empty
+   list of locations. The client can then reuse the existing
+   connection to the original server to fetch data. Otherwise, the
+   client must connect to one of the indicated locations.
+
+   In this way, the locations inside an endpoint can also be thought
+   of as performing lookaside load balacing or service discovery
```
Comment (wjones127) [inline docs/source/format/Flight.rst:113]:
```suggestion
of as performing look-aside load balancing or service discovery
```

**Review by wjones127 — COMMENTED** [review]
These diagrams are very helpful!

I've left a few suggestions.

File: docs/source/cpp/flight.rst (hunk 2 above)
Comment (lidavidm) [reply docs/source/cpp/flight.rst]: Good point. I removed/reworded things (except for the server-side RPC handlers where they still use out parameters).

**Review by wjones127 — APPROVED** [review]

**Pull request closed without merging**